src/coreclr/jit/runtimelookup.cpp

```diff
@@ -0,0 +1,461 @@
+                    }
+
+                    if (runtimeLookup.offsets[i] != 0)
+                    {
+                        if (isLastIndirectionWithSizeCheck)
+                        {
+                            lastIndOfTree = spillExpr(slotPtrTree);
+                            slotPtrTree   = gtClone(lastIndOfTree);
+                        }
+
+                        slotPtrTree = gtNewOperNode(GT_ADD, TYP_I_IMPL, slotPtrTree,
+                                                    gtNewIconNode(runtimeLookup.offsets[i], TYP_I_IMPL));
+                    }
+                }
+
+                // Non-dynamic expansion case (no size check):
+                //
+                // prevBb(BBJ_NONE):                    [weight: 1.0]
+                //     ...
+                //
+                // nullcheckBb(BBJ_COND):               [weight: 1.0]
+                //     if (*fastPathValue == null)
+                //         goto fallbackBb;
+                //
+                // fastPathBb(BBJ_ALWAYS):              [weight: 0.8]
+                //     rtLookupLcl = *fastPathValue;
+                //     goto block;
+                //
+                // fallbackBb(BBJ_NONE):                [weight: 0.2]
+                //     rtLookupLcl = HelperCall();
+                //
+                // block(...):                          [weight: 1.0]
+                //     use(rtLookupLcl);
+                //
+
+                // null-check basic block
+                BasicBlock* nullcheckBb = fgNewBBafter(BBJ_COND, prevBb, true);
+                nullcheckBb->bbFlags |= BBF_INTERNAL;
+
+                GenTree* fastPathValue = gtNewOperNode(GT_IND, TYP_I_IMPL, gtCloneExpr(slotPtrTree));
+                fastPathValue->gtFlags |= GTF_IND_NONFAULTING;
+
+                GenTree* fastPathValueClone =
+                    opts.OptimizationEnabled() ? fgMakeMultiUse(&fastPathValue) : gtCloneExpr(fastPathValue);
+
+                // Save dictionary slot to a local (to be used by fast path)
+                GenTree* nullcheckOp = gtNewOperNode(GT_EQ, TYP_INT, fastPathValue, gtNewIconNode(0, TYP_I_IMPL));
+                nullcheckOp->gtFlags |= GTF_RELOP_JMP_USED;
+                gtSetEvalOrder(nullcheckOp);
+                Statement* nullcheckStmt = fgNewStmtFromTree(gtNewOperNode(GT_JTRUE, TYP_VOID, nullcheckOp));
+                nullcheckStmt->SetDebugInfo(debugInfo);
+                gtSetStmtInfo(nullcheckStmt);
+                fgSetStmtSeq(nullcheckStmt);
+                fgInsertStmtAtEnd(nullcheckBb, nullcheckStmt);
+
+                // Fallback basic block
+                BasicBlock* fallbackBb = fgNewBBafter(BBJ_NONE, nullcheckBb, true);
+                fallbackBb->bbFlags |= BBF_INTERNAL;
+
+                GenTreeCall* fallbackCall = gtCloneExpr(call)->AsCall();
+                fallbackCall->gtArgs.GetArgByIndex(0)->SetLateNode(gtClone(ctxTree));
```

Comment:
This is relying on platform specific calling convention details. It would be better to reuse the original call directly or to create a new call from scratch and send it through call morphing.